pdf/pdfannotation.cpp

```diff
@@ -0,0 +1,356 @@
+
+PDFDocument* PDFAnnotation::document() const
+{
+    return d->m_document;
+}
+
+int PDFAnnotation::page() const
+{
+    return d->m_page;
+}
+
+QRectF PDFAnnotation::boundary() const
+{
+    return d->m_annotation->boundary();
+}
+
+void PDFAnnotation::attachOnce(PDFDocument *document, int page)
+{
+    /* Attach annotation only once. */
+    if (d->m_document != nullptr || document == nullptr)
+        return;
+    if (d->m_page >= 0 || page < 0 || page >= document->pageCount())
+        return;
+
+    d->m_document = document;
+    d->m_page = page;
+}
+
+void PDFAnnotation::attach(PDFDocument *document, PDFSelection *selection)
+{
+    if (!selection)
+        return;
+
+    QPair<int, QRectF> word = selection->rectAt(0);
+    attachOnce(document, word.first);
+    
+    QRectF bounds = word.second;
+    for (int i = 1; i < selection->count(); i++)
+        bounds.united(selection->rectAt(i).second);
+
+    d->m_annotation->setBoundary(bounds);
+    d->m_document->addAnnotation(d->m_annotation, d->m_page);
+    emit attached();
+}
+
+void PDFAnnotation::remove()
+{
+    if (d->m_document == nullptr)
+        return;
+
+    d->m_document->removeAnnotation(d->m_annotation, d->m_page);
+    d->m_annotation = nullptr;
+
+    deleteLater();
+}
+
+QString PDFAnnotation::author() const
+{
+    return d->m_annotation->author();
+}
+
+void PDFAnnotation::setAuthor(const QString &value)
+{
+    // Store last given author to use it as default for
+    // newly created annotations.
+    currentAuthor = value;
+
+    if (d->m_annotation->author() == value)
+        return;
+    d->m_annotation->setAuthor(value);
+    emit authorChanged();
+
+    d->m_annotation->setModificationDate(QDateTime());
+    emit modificationDateChanged();
+
+    if (d->m_document != nullptr)
+        d->m_document->documentModified();
+}
+
+QString PDFAnnotation::contents() const
+{
+    return d->m_annotation->contents();
+}
+
+void PDFAnnotation::setContents(const QString &value)
+{
+    if (d->m_annotation->contents() == value)
+        return;
+
+    d->m_annotation->setContents(value);
+    emit contentsChanged();
+
+    d->m_annotation->setModificationDate(QDateTime());
+    emit modificationDateChanged();
+
+    if (d->m_document != nullptr)
+        d->m_document->documentModified();
+}
+
+QDateTime PDFAnnotation::creationDate() const
+{
+    return d->m_annotation->creationDate();
+}
+
+QDateTime PDFAnnotation::modificationDate() const
+{
+    return d->m_annotation->modificationDate();
+}
+
+QColor PDFAnnotation::color() const
+{
+    return d->m_annotation->style().color();
+}
+
+void PDFAnnotation::setColor(const QColor &value)
+{
+    if (color() == value)
+        return;
+
+    Poppler::Annotation::Style style = d->m_annotation->style();
+    style.setColor(value);
+    d->m_annotation->setStyle(style);
+    emit colorChanged();
+
+    if (d->m_document != nullptr && d->m_page >= 0)
+        d->m_document->onPageChanged(d->m_page, d->m_annotation->boundary());
```

Comment:
Cannot modify page without touching the document it belongs to, right? Could then make this method trigger modification handling too and remove all documentModified() calls that are done afterwards.